### PR TITLE
functors_task_1_verna

### PR DESCRIPTION
**Pull request by VeeM**

### CODE REVIEW

File: functors_task_1/student/src/impl.cpp [hunk 1]
```diff
@@ -1,2 +1,44 @@
 #include "impl.h"
 
+#include <algorithm>
+#include <forward_list>
+#include <map>
+#include <utility>
+
+BeerBrand BeerOrganizer::operator()()
+{
+    if (m_nextBeer <= static_cast<BaseType>(BeerBrand::None)
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:10]: I think it is better to specify range explicitly with const variables: "to" and "from"

File: functors_task_1/student/src/impl.cpp [hunk 2]
```diff
@@ -1,2 +1,44 @@
...
+        m_nextBeer = static_cast<BaseType>(BeerBrand::None) + 1;
+    }
+
+    return static_cast<BeerBrand>(m_nextBeer++);
+}
+
+bool isExpensiveBeer(const BeerBrand brand)
+{
+    //TODO: move it somewhere
+    static std::forward_list<BeerBrand> expensiveBeer = {BeerBrand::Corona, BeerBrand::HoeGaarden};
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:22]: Wow, please implement this task without containers usage. Especially avoid lists (why - we'll talk about it at the second part of the training).

Don't overcomplicate here.

File: functors_task_1/student/src/impl.cpp [hunk 3]
```diff
@@ -1,2 +1,44 @@
...
+    return std::find(std::begin(expensiveBeer), std::end(expensiveBeer), brand) != std::end(expensiveBeer);
+}
+
+Cocktail MixingPolicy::mix(const AlcoholDrink lhd, const NonAlcoholDrink rhd)
+{
+    //TODO: move it somewhere
+    static std::map<std::pair<AlcoholDrink, NonAlcoholDrink>, Cocktail> recipes = {
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:29]: 1. Yep, mapping is a good pattern and should be used here, but please, implement this task without containers usage. You can easily replace it with a switch.
2. Why not const?

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: functors_task_1/student/src/impl.cpp (hunk 3 above)
Comment (smay1613) [reply functors_task_1/student/src/impl.cpp:29]: p.s also, you don't need ordered container here :)

File: functors_task_1/student/src/impl.cpp (hunk 2 above)
Comment (VeeM) [reply functors_task_1/student/src/impl.cpp:22]: Yes, the forward list shouldn't be used here.

And the one more thing. How I should guess that it is forbidden to use containers? :)

File: functors_task_1/student/src/impl.cpp (hunk 3 above)
Comment (VeeM) [reply functors_task_1/student/src/impl.cpp:29]: Should be const and unordered.
The same about containers as in the comment above nothing said about it in the task description. :)

File: functors_task_1/student/src/impl.cpp (hunk 1 above)
Comment (VeeM) [reply functors_task_1/student/src/impl.cpp:10]: You're right! Thanks!

File: functors_task_1/student/include/impl.h [hunk 4]
```diff
@@ -1,31 +1,41 @@
...
-{
-}
+bool isExpensiveBeer(const BeerBrand brand);
 
 /**
  * @todo Implement lambda beer country equality comparator
  * @return true if beer county is the same, false otherwise
  */
-auto sameCountry = [](/**???*/)
+auto sameCountry = [](BeerBrand lhd, BeerBrand rhd)
```
Comment (smay1613) [inline functors_task_1/student/include/impl.h:36]: _const here too :)_

File: functors_task_1/student/src/impl.cpp [hunk 5]
```diff
@@ -1,2 +1,53 @@
...
+            result = Cocktail::Greyhount;
+        }
+        break;
+    case AlcoholDrink::Whiskey :
+        if (NonAlcoholDrink::SevenUp == rhd)
+        {
+            result = Cocktail::SevenPlusSeven;
+        }
+        break;
+    case AlcoholDrink::Beer :
```
Comment (smay1613) [inline functors_task_1/student/src/impl.cpp:46]: _You can skip that_

File: functors_task_1/student/include/impl.h (hunk 4 above)
Comment (VeeM) [reply functors_task_1/student/include/impl.h:36]: О_о
Thanks :)

File: functors_task_1/student/src/impl.cpp (hunk 5 above)
Comment (VeeM) [reply functors_task_1/student/src/impl.cpp:46]: But in this case, we will get a warning(something like "not all enum entries was handled") ...

Comment (smay1613) [reply functors_task_1/student/src/impl.cpp:46]: Yep, and it is okay here, as for me

**Pull request closed without merging**